### PR TITLE
feat: add benchmarking suite

### LINKED ISSUE #9: [Testing] Enhance overall testing in the repo
## Objective

This sub-issue encompasses the changes needed to improve the overall testing in the repo. This should include:
 - Improved fuzzing
 - Benchmarking
 - Improvements to the current unit testing

## Origin Document

Issue #7 
[This comment](https://github.com/pokt-network/smt/pull/8#discussion_r1206094696)

## Goals

- Improve the fuzzing tests
  - Integrate with oss-fuzz for continuous fuzzing and integrate CI fuzzing tooling as appropriate
- Add benchmarking tests
  - Using numerous different inputs, benchmark individual components and grouped features of the SMT
- Improve current unit tests
  - Where it is appropriate cleanup, and improve the current test suite

## Deliverable

- [ ] Improve current fuzzing tests to fuzz numerous aspects of the SMT and its functionality
- [ ] Integrate with oss-fuzz for continuous fuzzing
- [ ] Add a CI workflow to be used on new PRs related to fuzzing
- [ ] Add benchmarking tests for components and feature flows of the SMT
- [ ] Improve current test suite

## Non-goals / Non-deliverables

- Change any underlying functionality of the SMT

## General issue deliverables

- [ ] Update any relevant README(s)
- [ ] Add or update any relevant or supporting [mermaid](https://mermaid-js.github.io/mermaid/) diagrams

## Testing Methodology

- [ ] **Task specific tests or benchmarks**: `go test ...`
- [ ] **New tests or benchmarks**: `go test ...`
- [ ] **All tests**: `go test -v`

---

**Creator**: @h5law
**Co-Owners**: @h5law

### PR DESCRIPTION
**Pull request by h5law**
### Summary generated by Reviewpad on 26 Sep 23 07:27 UTC
This pull request includes the following changes:

1. The Makefile has been modified to add several targets and their corresponding commands. The new targets are: help, list, test_all, benchmark_all, benchmark_smt, benchmark_smt_fill, benchmark_smt_ops, benchmark_smst, benchmark_smst_fill, and benchmark_smst_ops. Each target has a description comment above its definition.

2. The file diff adds two test functions: "TestSMT_ProofSizes" and "TestSMST_ProofSizes". These tests create a sparse Merkle tree and perform various operations on it. The tests calculate and log the average, minimum, and maximum sizes of serialized proofs and compacted proofs for different tree sizes ranging from 100,000 to 10,000,000.

3. The file "bench_test.go" has been deleted. It used to contain benchmark functions for updating and deleting data in a Sparse Merkle Tree. These benchmarks relied on a Key-Value Store and utilized the "testing" package. The functions performed operations such as updating and deleting data using the Sparse Merkle Tree. The benchmarking logic included timing and memory allocation measurements. The Key-Value Store was stopped at the end of each benchmark.

4. A new file called "bench_utils_test.go" has been introduced in the "benchmarks" package. The file contains benchmarking functions for the "smt" package and the "smst" package. It also includes helper functions for setting up the data structures and benchmarking functions for them. Overall, this file provides testing and benchmarking capabilities for the "smt" and "smst" packages.

5. Changes have been made to the ".gitignore" file to ignore the ".idea/" directory and remove the "testdata/" directory from being ignored.

6. The `.github/workflows/test.yml` file has been modified. Specifically, the `run` section of the `Create coverage report and run tests` job has been updated. Additionally, a new condition has been added to the `Sanitize test results` step.

7. The file "bench_leaf_test.go" has been modified to add two benchmark test functions: "BenchmarkSMTLeafSizes_Fill" and "BenchmarkSMSTLeafSizes_Fill". These benchmarks test the performance of filling a sparse Merkle tree with different sizes of leaves. They use different objects and parameters for the tree update operation.

Please let me know if you need further assistance with this diff!

## Issue

Fixes #9 

## Type of change

Please mark the relevant option(s):

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Major breaking change
- [x] Documentation
- [ ] Other 

## List of changes

- Add benchmarking suite for SM(S)T
- Add benchmark results to main readme

## Testing

- [x] **Task specific tests or benchmarks**: `go test ...`
- [x] **New tests or benchmarks**: `go test ...`
- [x] **All tests**: `go test -v`

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added, or updated, [`godoc` format comments](https://go.dev/blog/godoc) on touched members (see: [tip.golang.org/doc/comment](https://tip.golang.org/doc/comment))
- [x] I have tested my changes using the available tooling

### If Applicable Checklist

- [x] Update any relevant README(s)
- [ ] Add or update any relevant or supporting [mermaid](https://mermaid-js.github.io/mermaid/) diagrams
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: benchmarks/bench_leaf_test.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+	"crypto/sha256"
+	"fmt"
+	"github.com/pokt-network/smt"
+	"github.com/stretchr/testify/require"
+	"strconv"
+	"testing"
+)
+
+func BenchmarkSMTLeafSizes_Fill(b *testing.B) {
+	treeSizes := []int{100000, 500000, 1000000, 5000000, 10000000}
```
Comment (Olshansk) [inline benchmarks/bench_leaf_test.go:13]:
```suggestion
	treeSizes := []int{100000, 500000, 1000000, 5000000, 10000000} // num leaves
```

File: benchmarks/bench_leaf_test.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	leafSizes := []int{256, 512, 1024, 2048, 4096, 8192, 16384}
```
Comment (Olshansk) [inline benchmarks/bench_leaf_test.go:14]:
```suggestion
	leafSizes := []int{256, 512, 1024, 2048, 4096, 8192, 16384} // num bytes per leaf
```

File: benchmarks/bench_leaf_test.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	nodes, err := smt.NewKVStore("")
+	require.NoError(b, err)
+	for _, treeSize := range treeSizes {
+		for _, leafSize := range leafSizes {
+			leaf := make([]byte, leafSize)
+			for i := 0; i < 2; i++ {
+				commit := i == 1
+				name := "Fill"
```
Comment (Olshansk) [inline benchmarks/bench_leaf_test.go:22]: s/name/operation

File: benchmarks/bench_leaf_test.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+				b.ResetTimer()
+				b.Run(
+					fmt.Sprintf("%s [Leaf Size: %d bytes] (%d)", name, leafSize, treeSize),
+					func(b *testing.B) {
+						b.ResetTimer()
+						b.ReportAllocs()
+						for i := 0; i < treeSize; i++ {
+							require.NoError(b, tree.Update([]byte(strconv.Itoa(i)), leaf))
+						}
+						if commit {
```
Comment (Olshansk) [inline benchmarks/bench_leaf_test.go:36]: Do we have a test where you commit after every update and compare that?

Don't do it in this commit but leave a comment saying: `// There is an opportunity to do a fuzz test where we commit every N updates, if this ever becomes a bottlneck`

File: benchmarks/bench_leaf_test.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+func BenchmarkSMTLeafSizes_Fill(b *testing.B) {
+	treeSizes := []int{100000, 500000, 1000000, 5000000, 10000000}
+	leafSizes := []int{256, 512, 1024, 2048, 4096, 8192, 16384}
+	nodes, err := smt.NewKVStore("")
+	require.NoError(b, err)
+	for _, treeSize := range treeSizes {
+		for _, leafSize := range leafSizes {
+			leaf := make([]byte, leafSize)
+			for i := 0; i < 2; i++ {
```
Comment (Olshansk) [inline benchmarks/bench_leaf_test.go:20]: NIT: `for _, operation := range []string{"Commit", "Commit & Fill"} {`

File: benchmarks/bench_leaf_test.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+					},
+				)
+				require.NoError(b, nodes.ClearAll())
+			}
+		}
+	}
+	require.NoError(b, nodes.Stop())
+}
+
+func BenchmarkSMSTLeafSizes_Fill(b *testing.B) {
```
Comment (Olshansk) [inline benchmarks/bench_leaf_test.go:48]: Ditto everything above

File: benchmarks/bench_smt_test.go [hunk 7]
```diff
@@ -0,0 +1,385 @@
+package smt
```
Comment (Olshansk) [inline benchmarks/bench_smt_test.go:1]: Dude, great job on these tests!

Finding the balance of copy-paste vs abstraction without going overboard either way isn't easy, but you fucking nailed it!

File: benchmarks/bench_utils_test.go [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+		return err
+	}
+	delSMST = func(s *smt.SMST, i uint64) error {
+		b := make([]byte, 8)
+		binary.BigEndian.PutUint64(b, i)
+		return s.Delete(b)
+	}
+)
+
+func setupSMT(b *testing.B, persistent bool, num int) *smt.SMT {
```
Comment (Olshansk) [inline benchmarks/bench_utils_test.go:54]: NIT: Use a more semantic name than `num`

s/num/numLeaves

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: README.md [hunk 9]
```diff
@@ -19,4 +26,182 @@ This is a Go library that implements a Sparse Merkle tree for a key-value map. T
 
 Documentation for the different aspects of this library can be found in the [docs](./docs/) directory.
 
+## Benchmarks
+
+Benchmarks for the different aspects of this SMT library can be found in [benchmarks](./benchmarks/). In order to run the entire benchmarking suite use the following command:
+
+```sh
+go test -benchmem -run=^$ -bench Benchmark ./benchmarks -timeout 0
```
[the remark this answers is not in the document]
Comment (h5law) [reply README.md:34]: Yeah sure will do it in this PR I was thinking about it as well

[the remark this answers is not in the document]
Comment (Olshansk) [reply README.md:34]: Add TODOs for stuff like this

**Review by Olshansk — APPROVED** [review]
Left a couple more comments but feel free to merge after.

**Pull request merged**